### ISSUE TITLE
Support for AWS Elasticache Autodiscovery

### ISSUE THREAD

**Issue by marstonstudio** [killme2008/xmemcached#37]
Request to add support for AWS Autodiscovery to xmemcached
http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.html

Amazon forked the spy client to provide and implementation of this functionality
https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java/commit/70bf7643963500db20749d97c071b64b954eabb3

But we have had recurring problems with Spy client and much prefer xmemcached

Only problem with xmemcached is that there is no builtin recovery mechanism if a node is replaced after the client has initialized. A new node will have the same URL but different IP address.

Any interest?

**Comment by killme2008** [comment]
It would be a great feature.
Give me some time to read the document and commit.
If it's not complicated, i will try to implement it  ASAP,thanks.


2015-10-28 4:28 GMT+08:00 Jonathan Marston <notifications@github.com>:

> Request to add support for AWS Autodiscovery to xmemcached
>
> http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.html
>
> Amazon forked the spy client to provide and implementation of this
> functionality
> awslabs/aws-elasticache-cluster-client-memcached-for-java@70bf764
> <https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java/commit/70bf7643963500db20749d97c071b64b954eabb3>
>
> But we have had recurring problems with Spy client and much prefer
> xmemcached
>
> Only problem with xmemcached is that there is no builtin recovery
> mechanism if a node is replaced after the client has initialized. A new
> node will have the same URL but different IP address.
>
> Any interest?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/killme2008/xmemcached/issues/37>.
>



-- 
庄晓丹
Email:        killme2008@gmail.com xzhuang@avos.com
Site:           http://fnil.net
Twitter:      @killme2008

**Comment by flozano** [comment]
+1 to this issue, it's a big problem when Amazon replaces a node

**Comment by flozano** [comment]
Any update on this topic @killme2008 ? it seems we're seeing more node replacement by AWS these days and it's becoming a problem for us

**Comment by killme2008** [comment]
I am start working on it. But forgive me that i don't have much time on it.Try my best.

**Comment by killme2008** [comment]
hi，it's fixed in 2.1.0
https://github.com/killme2008/xmemcached/issues/49

**Comment by flozano** [comment]
Hi!

Thanks, but elasticache auto-discovery is not exactly the same as #49 (but #49 is totally needed to work properly with elasticache - thanks for that!)

Elasticache provides an  auto-discovery api:
http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.HowAutoDiscoveryWorks.html

This feature is very interesting when increasing or decreasing the cluster size dynamically.

**Comment by killme2008** [comment]
Sorry, my misunderstood.
It seems that i have to read the commit carefully.

https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java/commit/70bf7643963500db20749d97c071b64b954eabb3


I should resolve this issue in one week,very sorry so late to look into it.

**Comment by flozano** [comment]
No need to apologize, your work and your willingness to listen to users are greatly appreciated!

**Comment by killme2008** [comment]
@flozano  I will release a beta version this weekend, maybe you want to try it.

It may has some bug especially when adding/replacing/removing nodes.

**Comment by killme2008** [comment]
Released 2.2.1-beta

https://github.com/killme2008/xmemcached/releases/tag/xmemcached-2.2.1-beta

A simple example:

```java
import java.net.InetSocketAddress;

import net.rubyeye.xmemcached.aws.AWSElasticCacheClient;
import net.rubyeye.xmemcached.aws.ClusterConfigration;
import net.rubyeye.xmemcached.command.BinaryCommandFactory;
import net.rubyeye.xmemcached.command.TextCommandFactory;


public class Test {
    public static void main(String args[]) throws Exception {
        AWSElasticCacheClient client =
	    new AWSElasticCacheClient(new InetSocketAddress("AWS ElasticCache node host",
							    11211));
        ClusterConfigration conf = client.getConfig();
        System.out.println(conf.toString());
        client.set("a", 0, 1);
        System.out.println((Object)client.get("a"));
    }
}

```

**Comment by killme2008** [comment]
@marstonstudio 

Have you test it? Feedback welcome!

**Comment by marstonstudio** [comment]
Working in our production environment now.

**Comment by killme2008** [comment]
@marstonstudio  Thanks, i will release new version this weekend.

**Comment by killme2008** [comment]
Released 2.3.0, but it may take sometime to be synchronized to maven central repository.

https://github.com/killme2008/xmemcached/releases/tag/xmemcached-2.3.0
